### PR TITLE
refactor: improve ngWalker by preventing an error when a class has no name

### LINKED ISSUE #778: Codelyzer doesn't support export default
I reedit [my comment](https://github.com/mgechev/codelyzer/pull/754#discussion_r258851555) form the PR #754.

If we create a simple class like that:

```ts
export default class { ... }
```

and we launch `ng lint`, we have a big big crash [here] (https://github.com/mgechev/codelyzer/blob/master/src/propertyDecoratorBase.ts#L41) and many other places !

### PR DESCRIPTION
**Pull request by wKoza**
closes #778

### CODE REVIEW

File: src/contextualLifecycleRule.ts [hunk 1]
```diff
@@ -45,20 +46,18 @@ export class Rule extends AbstractRule {
...
 
-  protected visitNgPipe(controller: ClassDeclaration, decorator: Decorator): void {
-    this.validateDecorator(controller, decorator, METADATA_TYPE_LIFECYCLE_MAPPER.Pipe);
-    super.visitNgPipe(controller, decorator);
+  visitNgPipe(metadata: PipeMetadata): void {
+    this.validateDecorator(metadata.controller, metadata.decorator, METADATA_TYPE_LIFECYCLE_MAPPER.Pipe);
+    super.visitNgPipe(metadata);
   }
 
   private validateDecorator(controller: ClassDeclaration, decorator: Decorator, allowedMethods: ReadonlySet<LifecycleMethodKeys>): void {
```
Comment (rafaelss95) [inline src/contextualLifecycleRule.ts]:
```suggestion
   private validateDecorator(metadata: PipeMetadata, allowedMethods: ReadonlySet<LifecycleMethodKeys>): void {
```

File: src/angular/metadata.ts [hunk 2]
```diff
@@ -44,3 +44,20 @@ export class ComponentMetadata extends DirectiveMetadata {
     super(controller, decorator, selector);
   }
 }
+
+export class PipeMetadata {
+  constructor(
+    public readonly controller: ts.ClassDeclaration,
+    public readonly decorator: ts.Decorator,
+    public readonly name?: string,
```
Comment (rafaelss95) [inline src/angular/metadata.ts]:
```suggestion
    readonly name: string,
```

File: src/contextualLifecycleRule.ts [hunk 3]
```diff
@@ -45,20 +46,18 @@ export class Rule extends AbstractRule {
...
-    const className = getClassName(controller);
-
-    if (!className) return;
```
Comment (rafaelss95) [inline src/contextualLifecycleRule.ts]: nit: extra new line forgotten.

File: src/noPipeImpureRule.ts [hunk 4]
```diff
@@ -31,13 +32,13 @@ export class Rule extends AbstractRule {
...
-    super.visitNgPipe(controller, decorator);
+  protected visitNgPipe(metadata: PipeMetadata): void {
+    this.validatePipe(metadata);
+    super.visitNgPipe(metadata);
   }
 
-  private validatePipe(controller: ClassDeclaration, decorator: Decorator): void {
-    const pureExpression = getDecoratorPropertyInitializer(decorator, 'pure');
+  private validatePipe(metadata: PipeMetadata): void {
+    const pureExpression = getDecoratorPropertyInitializer(metadata.decorator, 'pure');
```
Comment (rafaelss95) [inline src/noPipeImpureRule.ts]: Now, I guess we can just `metadata.pure`, right? Do you prefer to do this kind of "refactor" in a new PR?

File: src/noPipeImpureRule.ts [hunk 5]
```diff
@@ -46,9 +47,7 @@ export class ClassMetadataWalker extends NgWalker {
 
     if (!parentExpression || isNotFalseLiteral) return;
 
-    const className = getClassName(controller);
-
-    if (!className) return;
```
Comment (rafaelss95) [inline src/noPipeImpureRule.ts]: nit: extra new line forgotten.

File: src/angular/metadata.ts [hunk 6]
```diff
@@ -44,3 +44,20 @@ export class ComponentMetadata extends DirectiveMetadata {
...
+    public readonly pure?: string
```
[the remark this answers is not in the document]
Comment (wKoza) [reply src/angular/metadata.ts]: It's the syndrome of the thing that we add at the last moment... `ts.BooleanLiteral` seems more convenient

File: src/angular/metadataReader.ts [hunk 7]
```diff
@@ -55,6 +92,26 @@ export class MetadataReader {
     return new DirectiveMetadata(d, dec, selector);
   }
 
+  protected readPipeMetadata(d: ts.ClassDeclaration, dec: ts.Decorator): DirectiveMetadata {
+    const name = this.getDecoratorArgument(dec)
+      .bind(expr => getStringInitializerFromProperty('name', expr!.properties))
+      .fmap(initializer => initializer!.text);
+
+    const pure = this.getDecoratorArgument(dec)
+      .bind(expr => getStringInitializerFromProperty('pure', expr!.properties))
```
[the remark this answers is not in the document]
Comment (wKoza) [reply src/angular/metadataReader.ts]: refactored

File: src/contextualLifecycleRule.ts [hunk 8]
```diff
@@ -45,20 +46,18 @@ export class Rule extends AbstractRule {
 }
 
 class ContextualLifecycleWalker extends NgWalker {
-  protected visitNgInjectable(controller: ClassDeclaration, decorator: Decorator): void {
-    this.validateDecorator(controller, decorator, METADATA_TYPE_LIFECYCLE_MAPPER.Injectable);
-    super.visitNgInjectable(controller, decorator);
+  visitNgInjectable(metadata: InjectableMetadata): void {
```
[the remark this answers is not in the document]
Comment (wKoza) [reply src/contextualLifecycleRule.ts]: the modifier ? if you think about `controller, decorator` -> `metadata` , it's just more consistent.

File: src/angular/ngWalker.ts [hunk 9]
```diff
@@ -165,13 +157,15 @@ export class NgWalker extends Lint.RuleWalker {
...
 
-  protected visitNgModule(decorator: ts.Decorator) {}
+  protected visitClassDecorator(decorator: ts.Decorator) {}
+
+  protected visitNgModule(metadata: ModuleMetadata) {}
 
   protected visitNgDirective(metadata: DirectiveMetadata) {}
 
-  protected visitNgPipe(controller: ts.ClassDeclaration, decorator: ts.Decorator) {}
+  protected visitNgPipe(metadata: DirectiveMetadata) {}
```
[the remark this answers is not in the document]
Comment (wKoza) [reply src/angular/ngWalker.ts]: good catch

File: src/contextualLifecycleRule.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (wKoza) [reply src/contextualLifecycleRule.ts]: it comes from the merge of my previous branch... I'm going to fix that also in `noOutputRenameRule`.

File: src/angular/metadata.ts (hunk 2 above)
Comment (wKoza) [reply src/angular/metadata.ts]: Unfortunately, We should keep `?`.  `name` may be not present in the source

File: src/noPipeImpureRule.ts (hunk 4 above)
Comment (wKoza) [reply src/noPipeImpureRule.ts]: It's done

File: src/angular/metadata.ts [hunk 10]
```diff
@@ -25,22 +25,35 @@ export interface TemplateMetadata extends PropertyMetadata {
...
     super(controller, decorator, selector);
   }
 }
+
+export class PipeMetadata {
+  constructor(
+    readonly controller: ts.ClassDeclaration,
+    readonly decorator: ts.Decorator,
+    readonly name?: string,
+    readonly pure?: ts.BooleanLiteral
```
Comment (rafaelss95) [inline src/angular/metadata.ts]: Hmm.. I'm unsure about this, if we go for this path, we should change `name` and `selector`_s_ to be `StringLiteralLike`.

IMHO, we should use `boolean`.

File: src/angular/metadata.ts [hunk 11]
```diff
@@ -25,39 +25,35 @@ export interface TemplateMetadata extends PropertyMetadata {
...
 
 export class PipeMetadata {
   constructor(
-    public readonly controller: ts.ClassDeclaration,
-    public readonly decorator: ts.Decorator,
-    public readonly name?: string,
-    public readonly pure?: string
+    readonly controller: ts.ClassDeclaration,
+    readonly decorator: ts.Decorator,
+    readonly name?: string,
```
Comment (rafaelss95) [inline src/angular/metadata.ts]: Actually `Angular` doesn't allow you to create a `@Pipe` without `name`...

File: src/angular/metadata.ts [hunk 12]
```diff
@@ -25,22 +25,35 @@ export interface TemplateMetadata extends PropertyMetadata {
...
   ) {
     super(controller, decorator, selector);
   }
 }
+
+export class PipeMetadata {
+  constructor(
+    readonly controller: ts.ClassDeclaration,
+    readonly decorator: ts.Decorator,
+    readonly name?: string,
```
Comment (rafaelss95) [inline src/angular/metadata.ts]: Actually `Angular` doesn't allow you to create a `@Pipe` without `name`...

File: src/angular/metadata.ts (hunk 11 above)
Comment (wKoza) [reply src/angular/metadata.ts]: Yes, of course but it's a static analyse.

File: src/angular/metadata.ts (hunk 10 above)
Comment (wKoza) [reply src/angular/metadata.ts]: Your comment here `noPipeImpureRule.spec.ts' is possible due to this choice. Maybe, in a second time, can we change all the string.

File: src/angular/metadataReader.ts [hunk 13]
```diff
@@ -55,6 +92,24 @@ export class MetadataReader {
     return new DirectiveMetadata(d, dec, selector);
   }
 
+  protected readPipeMetadata(d: ts.ClassDeclaration, dec: ts.Decorator): DirectiveMetadata {
+    const nameExpression = getDecoratorPropertyInitializer(dec, 'name');
+    const name = nameExpression && isStringLiteralLike(nameExpression) ? nameExpression.text : undefined;
+
+    const pureExpression = getDecoratorPropertyInitializer(dec, 'pure');
+    const isBoolean = pureExpression && isBooleanLiteralLike(pureExpression) ? pureExpression.kind : ts.SyntaxKind.FalseKeyword;
```
[the remark this answers is not in the document]
Comment (wKoza) [reply src/angular/metadataReader.ts]: It's a mistake.
```
const isBoolean = pureExpression && isBooleanLiteralLike(pureExpression);
```

should be enough.

File: src/angular/metadataReader.ts [hunk 14]
```diff
@@ -55,6 +92,24 @@ export class MetadataReader {
     return new DirectiveMetadata(d, dec, selector);
   }
 
+  protected readPipeMetadata(d: ts.ClassDeclaration, dec: ts.Decorator): DirectiveMetadata {
+    const nameExpression = getDecoratorPropertyInitializer(dec, 'name');
+    const name = nameExpression && isStringLiteralLike(nameExpression) ? nameExpression.text : undefined;
+
+    const pureExpression = getDecoratorPropertyInitializer(dec, 'pure');
+    const isBoolean = pureExpression && isBooleanLiteralLike(pureExpression);
```
Comment (rafaelss95) [inline src/angular/metadataReader.ts]:
```suggestion
    const pure = pureExpression && isBooleanLiteralLike(pureExpression) ? pureExpression : undefined;
```

This way you don't need to `cast` below.

File: src/angular/metadataReader.ts [hunk 15]
```diff
@@ -55,6 +92,24 @@ export class MetadataReader {
...
+
+    return new PipeMetadata(d, dec, name, isBoolean ? (pureExpression as ts.BooleanLiteral) : undefined);
```
Comment (rafaelss95) [inline src/angular/metadataReader.ts]:
```suggestion
    return new PipeMetadata(d, dec, name, pure);
```

File: src/noPipeImpureRule.ts [hunk 16]
```diff
@@ -31,27 +32,19 @@ export class Rule extends AbstractRule {
...
-    super.visitNgPipe(controller, decorator);
+  protected visitNgPipe(metadata: PipeMetadata): void {
+    this.validatePipe(metadata);
+    super.visitNgPipe(metadata);
   }
 
-  private validatePipe(controller: ClassDeclaration, decorator: Decorator): void {
-    const pureExpression = getDecoratorPropertyInitializer(decorator, 'pure');
+  private validatePipe(metadata: PipeMetadata): void {
+    if (!metadata.pure) return;
```
Comment (rafaelss95) [inline src/noPipeImpureRule.ts]:
```suggestion
    if (!metadata.pure || metadata.pure.kind !== SyntaxKind.FalseKeyword) return;
```

File: src/noPipeImpureRule.ts [hunk 17]
```diff
@@ -31,27 +32,19 @@ export class Rule extends AbstractRule {
...
+    if (metadata.pure!.kind !== SyntaxKind.FalseKeyword) return;
```
Comment (rafaelss95) [inline src/noPipeImpureRule.ts]:
```suggestion

```

**Comment by rafaelss95** [comment]
Some left minor comments and LGTM.

File: src/noPipeImpureRule.ts [hunk 18]
```diff
@@ -31,27 +32,18 @@ export class Rule extends AbstractRule {
...
-    super.visitNgPipe(controller, decorator);
+  protected visitNgPipe(metadata: PipeMetadata): void {
+    this.validatePipe(metadata);
+    super.visitNgPipe(metadata);
   }
 
-  private validatePipe(controller: ClassDeclaration, decorator: Decorator): void {
-    const pureExpression = getDecoratorPropertyInitializer(decorator, 'pure');
+  private validatePipe(metadata: PipeMetadata): void {
+    if (!metadata.pure || metadata.pure!.kind !== SyntaxKind.FalseKeyword) return;
```
Comment (rafaelss95) [inline src/noPipeImpureRule.ts]:
```suggestion
    if (!metadata.pure || metadata.pure.kind !== SyntaxKind.FalseKeyword) return;
```

Comment (rafaelss95) [reply src/noPipeImpureRule.ts]: You just missed the non null assertion 👀

Comment (wKoza) [reply src/noPipeImpureRule.ts]: arff

**Comment by wKoza** [comment]
ping @mgechev :tennis:

**Pull request merged**